### PR TITLE
Add support for HMS

### PR DESCRIPTION
**Pull request by k-jiang**
This PR add support for [Huawei Mobile Services](https://developer.huawei.com/consumer/en/doc/development/HMSCore-References-V5/api-common-statement-0000001050986127-V5) (HMS). It supports order verifications, server-to-server notification and purchase adjustments.

Code is modified based on [the HMS official demo](https://github.com/HMS-Core/hms-iap-serverdemo/tree/demo_golang).

Methods and models that included in this PR:

- Verification
    - Signature on user-provided info
    - Current subscriptions informations
    - Obtain canceled and / or refunded purchases
- Purchase adjustments
    - Cancel subscription auto-renewal
    - Manually extend a subscription
    - Issue refund on a subscriptions
- Models
    - InAppPurchaseData - contains information about a single subscription and order purchase
    - CanceledPurchase - information about purchase that is canceled by users or API calls.
    - StatusUpdateNotification - information when server-to-server notification is sent from HMS server, which is usually happened when a subscription changes status.

Please feel free to left comments below.

### CODE REVIEW

File: hms/validator.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+	if err := c.getResponseErrorByCode(resp.ResponseCode); err != nil {
+		return "", err
+	}
+
+	return resp.InappPurchaseData, nil
+}
+
+// OrderVerifyResponse JSON response from {rootUrl}/applications/purchases/tokens/verify
+type OrderVerifyResponse struct {
+	ResponseCode      string `josn:"responseCode"`                // Response code, if = "0" means succeed, for other see https://developer.huawei.com/consumer/en/doc/HMSCore-References-V5/server-error-code-0000001050166248-V5
```
Comment (takecy) [inline hms/validator.go:95]: It's a typo 🙈

File: hms/validator.go [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+	signatureByte, err := base64.StdEncoding.DecodeString(signature)
+	if err != nil {
+		return err
+	}
+	return rsa.VerifyPKCS1v15(pub.(*rsa.PublicKey), crypto.SHA256, hashed[:], signatureByte)
+}
+
+// SubscriptionVerifyResponse JSON response after requested {rootUrl}/sub/applications/v2/purchases/get
+type SubscriptionVerifyResponse struct {
+	ResponseCode      string `josn:"responseCode"`                // Response code, if = "0" means succeed, other means failed, for other see https://developer.huawei.com/consumer/en/doc/HMSCore-References-V5/server-error-code-0000001050166248-V5
```
Comment (takecy) [inline hms/validator.go:41]: It's a typo 🙈

**Review by takecy — CHANGES_REQUESTED** [review]
Thanks for the pull request!
Please correct the small typo,
and reabse to aggregate commits :)

**Comment by k-jiang** [comment]
Thank you for point it out. I have fix the typos.

As for the rebase, sure I can do it if it works out for you, but you can always chose "Squash and merge" if you want to keep the commit history tide. Please let me know how you want it be done.

image

**Comment by takecy** [comment]
Oh sorry, it wasn't rebase.
I would like to organize a commit history into just `add support for HMS`,
so please reset and force push to your master branch for unify those commits.
It's OK for the merge commits to be created.

**Comment by k-jiang** [comment]
I have force push the commits. Please take a look.

**Review by takecy — APPROVED** [review]
Thanks for the response!

**Pull request merged**